pip_faster.py

```diff
@@ -301,66 +292,95 @@ def add_entry(self, entry):
     """
-    return patched(vars(req), {'url': None, 'satisfied_by': None})
+    from copy import copy
+    req = copy(req)
+    req.url = None
+    req.satisfied_by = None
+    return req
 
 
 def trace_requirements(requirements):
     """given an iterable of pip InstallRequirements,
     return the set of required packages, given their transitive requirements.
     """
+    requirements = tuple(pretty_req(r) for r in requirements)
     working_set = fresh_working_set()
 
     # breadth-first traversal:
     from collections import deque
     queue = deque(requirements)
-    processed_reqs = set()
-    errors = set()
+    seen = set()
+    errors = []
     result = []
     while queue:
         req = queue.popleft()
-        processed_reqs.add(req.name)
+        req_id = str(req)  # InstallRequirement doesn't implement hash/equality, so we compare using str()
+        if req_id in seen:
+            logger.debug('already analyzed: %s', req)
+            continue
+        else:
+            seen.add(req_id)
+
         logger.debug('tracing: %s', req)
+        if req_cycle(req):
+            logger.warn('Circular dependency! %s', req)
+            continue
 
         try:
             dist = working_set.find(req.req)
         except pkg_resources.VersionConflict as conflict:
             dist = conflict.args[0]
-            with pretty_req(req):
-                errors.add('Error: version conflict: %s (%s) <-> %s' % (
-                    dist, timid_relpath(dist.location), req
-                ))
+            errors.append('Error: version conflict: %s (%s) <-> %s' % (
+                dist, timid_relpath(dist.location), req
+            ))
 
         if dist is None:
-            errors.add('Error: unmet dependency: %s' % req)
+            errors.append('Error: unmet dependency: %s' % req)
             continue
 
         result.append(dist_to_req(dist))
 
         for dist_req in sorted(dist.requires(), key=lambda req: req.key):
-            # temporarily shorten the str(req)
-            with pretty_req(req):
-                from pip.req import InstallRequirement
-                dist_req = InstallRequirement(dist_req, str(req))
+            from pip.req import InstallRequirement
+            dist_req = InstallRequirement(dist_req, req)
 
-            # in case there are circular dependencies, only process a
-            # requirement once
-            if dist_req.name not in processed_reqs:
-                logger.debug('adding sub-requirement %s', dist_req)
-                queue.append(dist_req)
+            logger.debug('adding sub-requirement %s', dist_req)
+            queue.append(dist_req)
 
     if errors:
-        raise InstallationError('\n'.join(sorted(errors)))
+        raise InstallationError('\n'.join(errors))
 
     return result
 
 
-def reqnames(reqs):
-    return set(req.name for req in reqs)
+class CacheOpts(object):
+
+    def __init__(self):
+        # We put the cache in the directory that pip already uses.
```

Comment:
The --find-links for pip-slower in venv-update during bootstrap was causing a warning in some cases, and wasn't really valuable, so I removed it. This means that venv-update doesn't use this object, so I moved it down to pip-faster.